### PR TITLE
Cimpl6 export

### PR DESCRIPTION
**Pull request by AmartC**
Updating the grammar to support CIMPL 6.0.

### CODE REVIEW

File: src/main/antlr/SHRDataElementLexer.g4 [hunk 1]
```diff
@@ -10,30 +10,30 @@ KW_USES:            'Uses:';
...
-KW_IS_TYPE:         'is type';
-KW_VALUE_IS_TYPE:   'value is type';
+KW_SUBSTITUTE:      'substitute';
 KW_INCLUDES:        'includes';
 KW_TRUE:            'true';
 KW_FALSE:           'false';
 KW_TBD:             'TBD';
 KW_TBD_CODE:        'TBD#TBD';
+KW_ONLY:            'only';
+KW_CODED_CONCEPT:   'concept';
```
Comment (cmoesel) [inline src/main/antlr/SHRDataElementLexer.g4]: The primitive lexer grammar already defines a token for `'concept'` and calls it `KW_CONCEPT_CODE`.  I don't think we need to define another token for it.  We should just use the primitive `KW_CONCEPT_CODE` since it is exactly the same -- and we're already importing the primitive lexer at the top of this file.

File: src/main/antlr/SHRDataElementLexer.tokens [hunk 2]
```diff
@@ -70,31 +64,32 @@ WS=69
 NEWLINE=70
 COMMENT=71
 LINE_COMMENT=72
+KW_PROPERTY=73
+PLUS=74
+KW_ONLY=75
+COLON=76
+OPEN_BRACKET=77
+CLOSE_BRACKET=78
+KW_CODED_CONCEPT=79
```
Comment (cmoesel) [inline src/main/antlr/SHRDataElementLexer.tokens]: Again, I think we can do without `KW_CODED_CONCEPT` since we already define a `KW_CONCEPT_CODE` in the primitive lexer.

File: src/main/antlr/SHRDataElementParser.g4 [hunk 3]
```diff
@@ -16,26 +16,32 @@ vocabularyDefs:     vocabularyDef+;
...
-values:             value? field*;
 
-value:              KW_VALUE count? OPEN_PAREN? valueType (KW_OR valueType)* CLOSE_PAREN?; // TODO: Remove PAREN (here for backwards compatibility now)
-valueType:          simpleOrFQName | ref | primitive | elementWithConstraint | tbd;
+field:              propertyField | elementWithConstraint;
 
-field:              count? OPEN_PAREN? fieldType (KW_OR fieldType)* CLOSE_PAREN?; // TODO: Remove PAREN (here for backwards compatibility now)
-fieldType:          specialWord | simpleOrFQName | ref | elementWithConstraint | tbd;
+propertyField:              KW_PROPERTY? propertyFieldType count;
+propertyFieldType:          specialWord | simpleOrFQName | tbd | KW_CODED_CONCEPT;
```
Comment (cmoesel) [inline src/main/antlr/SHRDataElementParser.g4]: Since `concept`is a primitive, I don't think we allow it to be declared as a `Property`.  We only allow primitives to be declared as a `Value` -- so you should remove ` | KW_CODED_CONCEPT`.

File: src/main/antlr/SHRDataElementParser.g4 [hunk 4]
```diff
@@ -16,26 +16,32 @@ vocabularyDefs:     vocabularyDef+;
...
 
-values:             value? field*;
 
-value:              KW_VALUE count? OPEN_PAREN? valueType (KW_OR valueType)* CLOSE_PAREN?; // TODO: Remove PAREN (here for backwards compatibility now)
-valueType:          simpleOrFQName | ref | primitive | elementWithConstraint | tbd;
+field:              propertyField | elementWithConstraint;
 
-field:              count? OPEN_PAREN? fieldType (KW_OR fieldType)* CLOSE_PAREN?; // TODO: Remove PAREN (here for backwards compatibility now)
-fieldType:          specialWord | simpleOrFQName | ref | elementWithConstraint | tbd;
+propertyField:              KW_PROPERTY? propertyFieldType count;
```
Comment (cmoesel) [inline src/main/antlr/SHRDataElementParser.g4]: If this is really a _property_ field, then I don't think `KW_PROPERTY` should be optional.  If the `Property:` keyword isn't there, then this is really an `elementWithConstraint`.  I _think_ you should be able to safely remove the `?` after `KW_PROPERTY`.

File: src/main/antlr/SHRDataElementParser.g4 [hunk 5]
```diff
@@ -44,36 +50,37 @@ descriptionProp:    KW_DESCRIPTION STRING;
...
 elementPath:                (specialWord | simpleOrFQName) (((DOT simpleName)+ (DOT primitive)?) | ((DOT simpleName)* DOT primitive));
-elementConstraint:          elementCodeVSConstraint | elementCodeValueConstraint | elementIncludesCodeValueConstraint | elementBooleanConstraint | elementTypeConstraint | elementIncludesTypeConstraint | elementWithUnitsConstraint;
-legacyWithCode:             KW_WITH (KW_CODE | simpleOrFQName); // Just here for backwards compatibility until definitions are updated
-elementCodeVSConstraint:    legacyWithCode? bindingInfix? KW_FROM valueset KW_IF_COVERED?;
-elementCodeValueConstraint: KW_IS codeOrFQCode;
+//elementBracketPath:         (specialWord | simpleOrFQName) (((OPEN_BRACKET simpleName CLOSE_BRACKET)+ (DOT primitive)?) | ((OPEN_BRACKET simpleName CLOSE_BRACKET)* DOT primitive) | ((DOT simpleName)* OPEN_BRACKET (simpleName | primitive) CLOSE_BRACKET)| ((DOT simpleName)* DOT primitive));
+//elementAllBrackets:           (specialWord | simpleOrFQName) (((OPEN_BRACKET simpleName CLOSE_BRACKET)+ (OPEN_BRACKET primitive CLOSE_BRACKET)?) | ((OPEN_BRACKET simpleName CLOSE_BRACKET)* OPEN_BRACKET primitive CLOSE_BRACKET));
```
Comment (cmoesel) [inline src/main/antlr/SHRDataElementParser.g4]: Are the two commented lines above still needed?  If not, please delete them.

File: src/main/antlr/SHRDataElementParser.g4 [hunk 6]
```diff
@@ -44,36 +50,37 @@ descriptionProp:    KW_DESCRIPTION STRING;
...
+elementBracketPath:         (specialWord | simpleOrFQName) (OPEN_BRACKET (simpleName | primitive) CLOSE_BRACKET)* (DOT simpleName (OPEN_BRACKET (simpleName | primitive) CLOSE_BRACKET)*)*;
+elementConstraint:          elementCodeVSConstraint | elementCodeValueConstraint | elementIncludesCodeValueConstraint | elementBooleanConstraint | elementTypeConstraint | elementIncludesTypeConstraint | elementUrlConstraint;
+elementCodeVSConstraint:    (KW_CODED_CONCEPT | simpleOrFQName)? KW_FROM valueset (OPEN_PAREN bindingStrength CLOSE_PAREN)?;
```
Comment (cmoesel) [inline src/main/antlr/SHRDataElementParser.g4]: I don't think you should have `(KW_CODED_CONCEPT | simpleOrFQName)?` here.  This is supposed to just have the constraint phrase itself.  For example, notice that `elementCodeValueConstraint` is just `EQUAL codeOrFQCode`.  `elementCodeVSConstraint` should be similar.

File: src/main/antlr/SHRDataElementParser.g4 [hunk 7]
```diff
@@ -44,36 +50,37 @@ descriptionProp:    KW_DESCRIPTION STRING;
...
+elementTypeConstraint:      (KW_SUBSTITUTE | KW_ONLY) (simpleOrFQName | primitive | tbd);
+elementUrlConstraint:       EQUAL URL;
 elementIncludesTypeConstraint: (KW_INCLUDES typeConstraint)+;
-elementWithUnitsConstraint: KW_WITH KW_UNITS fullyQualifiedCode;
 valueset:           URL | PATH_URL | URN_OID | simpleName | tbd;
 primitive:          KW_BOOLEAN | KW_INTEGER | KW_STRING | KW_DECIMAL | KW_URI | KW_BASE64_BINARY | KW_INSTANT | KW_DATE
-                    | KW_DATE_TIME | KW_TIME | KW_CODE | KW_OID | KW_ID | KW_MARKDOWN | KW_UNSIGNED_INT
-                    | KW_POSITIVE_INT | KW_XHTML;
+                    | KW_DATE_TIME | KW_TIME | KW_CONCEPT_CODE | KW_OID | KW_ID | KW_MARKDOWN | KW_UNSIGNED_INT
+                    | KW_POSITIVE_INT | KW_XHTML | KW_CODED_CONCEPT;
```
Comment (cmoesel) [inline src/main/antlr/SHRDataElementParser.g4]: Again, if you remove the redundant `KW_CODED_CONCEPT` definition, then you can remove the reference to it here as well -- especially since you're already covering `KW_CONCEPT_CODE` in the line above.

**Review by cmoesel — CHANGES_REQUESTED** [review]
I've noted a few recommended changes in separate comments.  They're all pretty minor, so should be straight-forward.  In addition, please remove the checked in `.DS_Store` file.

Thanks!  It's almost there!

**Comment by cmoesel** [comment]
The fixes look good.  I ran the parser against the output of the CIMPL6 exporter.  Some of the VS files failed -- but I think that's an issue in the exporter, not the grammar.  In the exporter PR, I suggested _not_ outputting the map or VS files at all.

Only one dataelement file had a parsing error:
```
Parsing ../shr-cli/out/cimpl6/shr_core.txt
line 112:19 no viable alternative at input 'concept'
```
But... this is actually OK.  It's the definition of `CodeableConcept`:
```
Element:           CodeableConcept
Description:       "A set of codes drawn from different coding systems, representing the same concept."
Property:          concept 0..*
Property:          DisplayText 0..1
```
The error happens because `concept` cannot be a property (since it is a primitive).  So the grammar is working correctly.  And it's an edge case in the exporter.  That property should actually be `Coding`, but since we replace all `Coding` w/ `concept`, that causes this problem.  I think this fits into the 5% of the exporter that we should just fix by hand after export.  In fact, I think we plan to _remove_ `CodeableConcept` and `Coding` definitions anyway.

So, long story short (too late) -- this looks good!  I will approve!

**Review by cmoesel — APPROVED** [review]
